opts/mount.go

```diff
@@ -112,6 +113,33 @@ func (m *MountOpt) Set(value string) error {
 			if err != nil {
 				return fmt.Errorf("invalid value for %s: %s", key, val)
 			}
+			logrus.Warn("bind-nonrecursive is deprecated, use bind-recursive=disabled instead")
+		case "bind-recursive":
+			valS := val
+			// Allow boolean as an alias to "enabled" or "disabled"
+			if b, err := strconv.ParseBool(valS); err == nil {
+				if b {
+					valS = "enabled"
+				} else {
+					valS = "disabled"
+				}
+			}
+			switch valS {
+			case "enabled": // read-only mounts are recursively read-only if Engine >= v25 && kernel >= v5.12
+				// NOP
+			case "disabled": // alias of bind-nonrecursive=true
+				bindOptions().NonRecursive = true
+			case "readonly-classic": // conforms to Docker v24; read-only mounts are recursively mounted but not recursively read-only
+				bindOptions().ReadOnlyNonRecursive = true
+				mount.ReadOnly = true
+			case "readonly": // force recursively read-only, or raise an error
+				bindOptions().ReadOnlyForceRecursive = true
+				bindOptions().Propagation = mounttypes.PropagationRPrivate
```

Comment:
I don't think we should overwrite the propagation here unconditionally. I think it's fine to set it implicitly from a UX perspective, but if the user specifies any different propagation it should be a validation error.

Reply:
Updated

Reply:
Hmm, I do like the trivial validation overall, but I'm not sure about this case. I don't think that requiring the user to set the propagation explicitly is good UX; I'd prefer this to be set implicitly, but to error if the user explicitly sets a different propagation. Unfortunately that doesn't work with the simple validation you've added, but maybe you have some ideas to that end?

Reply:
The default implicit "rprivate" propagation sometimes flips to "rslave":
- https://github.com/moby/moby/pull/36055

So I guess we should demand explicit "rprivate" here

Reply:
Yeah, I had that in mind. What I meant to say is that we want to be explicit at the API level, but I'm not sure that we want to require the user to explicitly specify a propagation mode to use `bind-recursive=readonly`. However with the current validation, this is messy.

I am okay with what you have; it makes the feature usable, and `bind-recursive=readonly` is definitely a power-user feature/we're "upgrading" users by default to the new `enabled` behavior anyway.

How about a `// TODO:` to implicitly set propagation and error if the user specifies a propagation in a future refactor/UX polish pass?

Reply:
Added